core/server/master/src/main/java/alluxio/master/file/MetadataSyncLockManager.java

```diff
@@ -0,0 +1,110 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file;
+
+import alluxio.AlluxioURI;
+import alluxio.collections.LockPool;
+import alluxio.concurrent.LockMode;
+import alluxio.conf.Configuration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.InvalidPathException;
+import alluxio.resource.LockResource;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Class for managing metadata　sync locking.
+ */
+public class MetadataSyncLockManager {
+  /**
+   * Pool for supplying metadata sync locks.
+   *
+   * We use weak values so that when nothing holds a reference to
+   * a lock, the garbage collector can remove the lock's entry from the pool.
+   */
+  private final LockPool<String> mLockPool =
+      new LockPool<>((key) -> new ReentrantReadWriteLock(),
+          Configuration.getInt(PropertyKey.MASTER_METADATA_SYNC_LOCK_POOL_INITSIZE),
+          Configuration.getInt(PropertyKey.MASTER_METADATA_SYNC_LOCK_POOL_LOW_WATERMARK),
+          Configuration.getInt(PropertyKey.MASTER_METADATA_SYNC_LOCK_POOL_HIGH_WATERMARK),
+          Configuration.getInt(PropertyKey.MASTER_METADATA_SYNC_LOCK_POOL_CONCURRENCY_LEVEL));
+
+  /**
+   * Acquire locks for a given path before metadata sync.
+   * Locks are for leading paths of a uri instead of inodes
+   * and hence paths will be locked regardless if they exist in alluxio or not.
+   * All locks but the last one will be read locks and the last lock acquired will be a write lock.
+   *
+   * For example, for path /a/b/c/d, this method will acquire the following locks sequentially
+   * 1. a read lock on /
+   * 2. a read lock on /a/
+   * 3. a read lock on /a/b/
+   * 4. a read lock on /a/b/c/
+   * 5. a write lock on /a/b/c/d/
+   *
+   * @param uri the alluxio uri to perform metadata sync
+   * @return the {@link MetadataSyncPathList} representing the locked paths
+   */
+  public MetadataSyncPathList lockPath(AlluxioURI uri) throws InvalidPathException, IOException {
+    MetadataSyncPathList list = new MetadataSyncPathList();
+    String[] components = PathUtils.getPathComponents(uri.getPath());
+    StringBuilder sb = new StringBuilder();
+    // TODO(elega) potential inefficient string operations
+    try {
+      for (int i = 0; i < components.length; ++i) {
+        sb.append(components[i]);
+        sb.append('/');
+        String lockKey = sb.toString();
+        Preconditions.checkNotNull(lockKey);
+        LockMode lockMode = (i == components.length - 1) ? LockMode.WRITE : LockMode.READ;
+        list.add(mLockPool.get(lockKey, lockMode));
+      }
+    } catch (Throwable t) {
```

Comment:
However this should matter little, because we are mostly using daemon threads in the JVM so even if you should interrupt a thread but you didn't, that thread doesn't block the jvm from exiting. Not sure what else we use interruption for.

Reply:
InterruptedException is a checked exception and there is no call in the method would throw such a exception so no need to handle it.